File: download_wiki.py
#!/usr/bin/env python3
"""Download BitLife Fandom wiki pages through the MediaWiki API.

The script stores one JSON file per page plus a manifest. It uses only the
Python standard library so it can run in the bundled Codex runtime.
"""

from __future__ import annotations

import argparse
import json
import os
import re
import time
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Any, Iterable


API_ENDPOINT = "https://bitlife-life-simulator.fandom.com/api.php"
USER_AGENT = "bitliffe-wiki-downloader/1.0 (personal research)"


def safe_filename(value: str) -> str:
    value = value.replace(" ", "_")
    value = re.sub(r"[^A-Za-z0-9_\-\.]+", "", value)
    return value[:180] or "untitled"


def api_get(params: dict[str, Any], retries: int = 3) -> dict[str, Any]:
    encoded = urllib.parse.urlencode(params, doseq=True)
    request = urllib.request.Request(
        f"{API_ENDPOINT}?{encoded}",
        headers={
            "Accept": "application/json",
            "Referer": "https://bitlife-life-simulator.fandom.com/",
            "User-Agent": USER_AGENT,
        },
    )

    last_error: Exception | None = None
    for attempt in range(1, retries + 1):
        try:
            with urllib.request.urlopen(request, timeout=30) as response:
                body = response.read().decode("utf-8")
                return json.loads(body)
        except Exception as exc:  # noqa: BLE001 - keep retries broad for network flakes.
            last_error = exc
            if attempt == retries:
                break
            time.sleep(1.5 * attempt)
    raise RuntimeError(f"API request failed after {retries} attempts: {last_error}")


def get_namespaces() -> list[int]:
    data = api_get(
        {
            "action": "query",
            "format": "json",
            "meta": "siteinfo",
            "siprop": "namespaces",
        }
    )
    namespaces = data.get("query", {}).get("namespaces", {})
    return sorted(int(ns) for ns in namespaces if int(ns) >= 0)


def page_filename(page: dict[str, Any]) -> str:
    return f"{page.get('pageid', 'missing')}_{safe_filename(page.get('title', 'untitled'))}.json"


def extract_content(page: dict[str, Any]) -> str:
    revisions = page.get("revisions") or []
    if not revisions:
        return ""
    revision = revisions[0]
    slots = revision.get("slots") or {}
    main = slots.get("main") or {}
    return main.get("content") or main.get("*") or revision.get("*") or ""


def iter_pages(namespace: int, limit: int = 0) -> Iterable[dict[str, Any]]:
    params: dict[str, Any] = {
        "action": "query",
        "format": "json",
        "formatversion": "2",
        "generator": "allpages",
        "gaplimit": "50",
        "gapnamespace": namespace,
        "prop": "revisions",
        "rvprop": "ids|timestamp|user|content",
        "rvslots": "main",
    }
    yielded = 0

    while True:
        data = api_get(params)
        for page in data.get("query", {}).get("pages", []):
            page["content"] = extract_content(page)
            yield page
            yielded += 1
            if limit and yielded >= limit:
                return

        continuation = data.get("continue")
        if not continuation:
            return
        params.update(continuation)
        time.sleep(0.2)


def load_existing_manifest(output_dir: Path) -> list[dict[str, Any]]:
    manifest_path = output_dir / "manifest.json"
    if not manifest_path.exists() or manifest_path.stat().st_size == 0:
        return []
    try:
        data = json.loads(manifest_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return []
    return data if isinstance(data, list) else []


def write_manifest(output_dir: Path) -> None:
    pages_dir = output_dir / "pages"
    manifest: list[dict[str, Any]] = []
    for path in sorted(pages_dir.glob("*.json")):
        try:
            page = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            continue
        manifest.append(
            {
                "pageid": page.get("pageid"),
                "title": page.get("title"),
                "ns": page.get("ns"),
                "file": str(path.relative_to(output_dir)).replace(os.sep, "/"),
            }
        )
    (output_dir / "manifest.json").write_text(
        json.dumps(manifest, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )


def parse_namespaces(value: str, all_namespaces: bool) -> list[int]:
    if all_namespaces:
        return get_namespaces()
    return [int(part.strip()) for part in value.split(",") if part.strip()]


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output-dir", default="wiki_dump")
    parser.add_argument("--limit", type=int, default=0, help="Limit total pages (0 = no limit)")
    parser.add_argument(
        "--namespaces",
        default="0",
        help="Comma-separated MediaWiki namespaces to download. Default: 0 (main articles).",
    )
    parser.add_argument(
        "--all-namespaces",
        action="store_true",
        help="Download all non-negative namespaces, including files, templates, and categories.",
    )
    parser.add_argument(
        "--refresh",
        action="store_true",
        help="Overwrite existing page JSON files instead of skipping them.",
    )
    args = parser.parse_args()

    output_dir = Path(args.output_dir)
    pages_dir = output_dir / "pages"
    pages_dir.mkdir(parents=True, exist_ok=True)

    namespaces = parse_namespaces(args.namespaces, args.all_namespaces)
    existing = {item.get("pageid") for item in load_existing_manifest(output_dir)}
    saved = 0
    skipped = 0

    print(f"Downloading namespaces: {', '.join(map(str, namespaces))}")
    for namespace in namespaces:
        print(f"Namespace {namespace}: listing and fetching pages...")
        for page in iter_pages(namespace):
            if args.limit and saved >= args.limit:
                break
            pageid = page.get("pageid")
            path = pages_dir / page_filename(page)
            if not args.refresh and (pageid in existing or path.exists()):
                skipped += 1
                continue
            path.write_text(json.dumps(page, ensure_ascii=False, indent=2), encoding="utf-8")
            saved += 1
            if saved % 50 == 0:
                print(f"Saved {saved} new pages ({skipped} skipped)...")
        if args.limit and saved >= args.limit:
            break

    write_manifest(output_dir)
    manifest_count = len(load_existing_manifest(output_dir))
    print(f"Done. Saved {saved} new pages, skipped {skipped}, manifest has {manifest_count} pages.")


if __name__ == "__main__":
    main()

File: download_wiki.mjs
#!/usr/bin/env node

import fs from "node:fs/promises";
import path from "node:path";

const API_ENDPOINT = "https://bitlife-life-simulator.fandom.com/api.php";
const USER_AGENT = "bitliffe-wiki-downloader/1.0 (personal research)";

function getArg(name, fallback = undefined) {
  const index = process.argv.indexOf(name);
  if (index === -1) return fallback;
  return process.argv[index + 1] ?? fallback;
}

function hasFlag(name) {
  return process.argv.includes(name);
}

function safeFilename(value) {
  const safe = value.replaceAll(" ", "_").replace(/[^A-Za-z0-9_\-.]+/g, "").slice(0, 180);
  return safe || "untitled";
}

async function apiGet(params) {
  const url = new URL(API_ENDPOINT);
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, String(value));
  }

  let lastError;
  for (let attempt = 1; attempt <= 3; attempt += 1) {
    try {
      const response = await fetch(url, {
        headers: {
          accept: "application/json",
          referer: "https://bitlife-life-simulator.fandom.com/",
          "user-agent": USER_AGENT,
        },
      });
      const text = await response.text();
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${text.slice(0, 200)}`);
      }
      if (!text) {
        throw new Error("empty response body");
      }
      return JSON.parse(text);
    } catch (error) {
      lastError = error;
      if (attempt < 3) {
        await new Promise((resolve) => setTimeout(resolve, attempt * 1500));
      }
    }
  }
  throw lastError;
}

async function getNamespaces() {
  const data = await apiGet({
    action: "query",
    format: "json",
    meta: "siteinfo",
    siprop: "namespaces",
    origin: "*",
  });
  return Object.keys(data.query?.namespaces ?? {})
    .map((value) => Number(value))
    .filter((value) => value >= 0)
    .sort((a, b) => a - b);
}

function extractContent(page) {
  const revision = page.revisions?.[0];
  const main = revision?.slots?.main;
  return main?.content ?? main?.["*"] ?? revision?.["*"] ?? "";
}

async function writeManifest(outputDir, pagesDir) {
  const files = (await fs.readdir(pagesDir)).filter((file) => file.endsWith(".json")).sort();
  const manifest = [];
  for (const file of files) {
    try {
      const page = JSON.parse(await fs.readFile(path.join(pagesDir, file), "utf8"));
      manifest.push({
        pageid: page.pageid,
        title: page.title,
        ns: page.ns,
        file: `pages/${file}`,
      });
    } catch {
      // Ignore partial files from interrupted runs.
    }
  }
  await fs.writeFile(
    path.join(outputDir, "manifest.json"),
    `${JSON.stringify(manifest, null, 2)}\n`,
    "utf8",
  );
  return manifest.length;
}

async function downloadNamespace(namespace, pagesDir, options) {
  let continuation;
  let saved = 0;
  let skipped = 0;

  while (true) {
    const params = {
      action: "query",
      format: "json",
      formatversion: "2",
      generator: "allpages",
      gaplimit: "50",
      gapnamespace: namespace,
      prop: "revisions",
      rvprop: "ids|timestamp|user|content",
      rvslots: "main",
      origin: "*",
    };
    if (continuation) {
      params.gapcontinue = continuation;
    }

    const data = await apiGet(params);
    const pages = data.query?.pages ?? [];
    for (const page of pages) {
      if (options.limit > 0 && options.totalSaved >= options.limit) {
        return { saved, skipped, done: true };
      }

      const file = `${page.pageid}_${safeFilename(page.title)}.json`;
      const filePath = path.join(pagesDir, file);
      if (!options.refresh) {
        try {
          await fs.access(filePath);
          skipped += 1;
          continue;
        } catch {
          // File does not exist yet.
        }
      }

      const output = {
        pageid: page.pageid,
        title: page.title,
        ns: page.ns,
        touched: page.touched,
        lastrevid: page.lastrevid,
        content: extractContent(page),
      };
      await fs.writeFile(filePath, `${JSON.stringify(output, null, 2)}\n`, "utf8");
      saved += 1;
      options.totalSaved += 1;
      if (options.totalSaved % 50 === 0) {
        console.log(`Saved ${options.totalSaved} new pages...`);
      }
    }

    continuation = data.continue?.gapcontinue;
    if (!continuation) {
      return { saved, skipped, done: false };
    }
    await new Promise((resolve) => setTimeout(resolve, 200));
  }
}

async function main() {
  const outputDir = getArg("--output-dir", "wiki_dump");
  const limit = Number(getArg("--limit", "0"));
  const namespaceArg = getArg("--namespaces", "0");
  const refresh = hasFlag("--refresh");
  const allNamespaces = hasFlag("--all-namespaces");
  const namespaces = allNamespaces
    ? await getNamespaces()
    : namespaceArg.split(",").map((value) => Number(value.trim())).filter(Number.isFinite);

  const pagesDir = path.join(outputDir, "pages");
  await fs.mkdir(pagesDir, { recursive: true });

  const options = { limit, refresh, totalSaved: 0 };
  let totalSkipped = 0;

  console.log(`Downloading namespaces: ${namespaces.join(", ")}`);
  for (const namespace of namespaces) {
    console.log(`Namespace ${namespace}: listing and fetching pages...`);
    const result = await downloadNamespace(namespace, pagesDir, options);
    totalSkipped += result.skipped;
    if (result.done) break;
  }

  const manifestCount = await writeManifest(outputDir, pagesDir);
  console.log(
    `Done. Saved ${options.totalSaved} new pages, skipped ${totalSkipped}, manifest has ${manifestCount} pages.`,
  );
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
